File: explain_analyze.R
# EXPLAIN ANALYZE of one wide LEFT JOIN from the workload (R client).
# Usage: Rscript explain_analyze.R [threads]   (default 8)
suppressWarnings(suppressPackageStartupMessages({library(DBI); library(duckdb)}))
args <- commandArgs(trailingOnly = TRUE)
threads <- if (length(args) >= 1) as.integer(args[1]) else 8L

con <- suppressMessages(dbConnect(duckdb(), dbdir = ":memory:"))
invisible(dbExecute(con, sprintf("SET threads TO %d", threads)))

payload <- paste(sprintf("'p%d_' || CAST(hash(i * %d) %% 100000 AS VARCHAR) AS p%d",
                         1:8, (1:8) * 7L + 3L, 1:8), collapse = ", ")
invisible(dbExecute(con, sprintf(
  "CREATE TABLE t AS SELECT i AS id, 'k' || CAST(hash(i) %% 2000000 AS VARCHAR) AS key, %s FROM range(8000000) g(i)", payload)))
invisible(dbExecute(con, "CREATE TABLE d AS SELECT 'k' || CAST(i AS VARCHAR) AS key, 'v' || CAST(i AS VARCHAR) AS value FROM range(2000000) g(i)"))
for (k in 1:6) invisible(dbExecute(con, sprintf(
  "CREATE OR REPLACE TABLE t AS SELECT t.*, d.value AS w%d FROM t LEFT JOIN d ON t.key = d.key", k)))

cat(sprintf("# platform=%s  duckdb=%s  threads=%d\n",
    dbGetQuery(con, "PRAGMA platform")[[1]], as.character(packageVersion("duckdb")), threads))
plan <- dbGetQuery(con,
  "EXPLAIN ANALYZE CREATE OR REPLACE TABLE t2 AS SELECT t.*, d.value AS w FROM t LEFT JOIN d ON t.key = d.key")
cat(plan[[ncol(plan)]], sep = "\n")
dbDisconnect(con, shutdown = TRUE)

File: explain_analyze.py
"""EXPLAIN ANALYZE of one wide LEFT JOIN from the workload (Python client).
Usage: python explain_analyze.py [threads]   (default 8)
Prints PRAGMA platform, duckdb version, and the profiled plan tree."""
import duckdb, sys
sys.stdout.reconfigure(encoding="utf-8")  # the plan tree uses box-drawing chars

threads = int(sys.argv[1]) if len(sys.argv) > 1 else 8
con = duckdb.connect(":memory:")
con.execute(f"SET threads TO {threads}")

payload = ", ".join(f"'p{j}_' || CAST(hash(i * {j*7+3}) % 100000 AS VARCHAR) AS p{j}" for j in range(1, 9))
con.execute(f"CREATE TABLE t AS SELECT i AS id, 'k' || CAST(hash(i) % 2000000 AS VARCHAR) AS key, {payload} FROM range(8000000) g(i)")
con.execute("CREATE TABLE d AS SELECT 'k' || CAST(i AS VARCHAR) AS key, 'v' || CAST(i AS VARCHAR) AS value FROM range(2000000) g(i)")
for k in range(1, 7):  # widen t to match a late join in the workload
    con.execute(f"CREATE OR REPLACE TABLE t AS SELECT t.*, d.value AS w{k} FROM t LEFT JOIN d ON t.key = d.key")

print(f"# platform={con.execute('PRAGMA platform').fetchone()[0]}  duckdb={duckdb.__version__}  threads={threads}")
plan = con.execute(
    "EXPLAIN ANALYZE CREATE OR REPLACE TABLE t2 AS "
    "SELECT t.*, d.value AS w FROM t LEFT JOIN d ON t.key = d.key"
).fetchall()[0][1]
print(plan)
con.close()
